### DISCUSSION TITLE
Class-Balancing Reservoir Sampling

### DISCUSSION THREAD

**Discussion by sarat2225** [ContinualAI/avalanche#955] [Feature Request]
Hey!
I am trying to implement CBRS Replay method as a plugin in Avalanche following this [article](http://proceedings.mlr.press/v119/chrysakis20a/chrysakis20a.pdf). Is there any previous work on this?

**Comment by AndreaCossu** [comment]
Not that I know of. In case they are helpful, you can start by taking a look at the [existing dataloders in Avalanche](https://avalanche-api.continualai.org/en/latest/benchmarks.html?highlight=groupbalanceddataloader#data-loaders).

**Comment by AntonioCarta** [comment]
*not familiar with the paper but I'm assuming they just do standard reservoir sampling + class-based splits.*

We have reservoir sampling and class-splitted buffers ([here](https://avalanche-api.continualai.org/en/v0.1.0/training.html#replay-buffers)). Maybe the easiest solution would be to use a `ParametricBuffer` with class splits and implement a selection strategy equivalent to the one used by the `ReservoirSamplingBuffer`.

**Reply by sarat2225** [reply]
Thanks for the simpler solution 😄. We will try it.

**Reply by sarat2225** [reply]
Hi!
We implemented the algorithm-2 in the above [article](http://proceedings.mlr.press/v119/chrysakis20a/chrysakis20a.pdf) but we want to pass custom loss function and the losses are computed in the 1st inner for  loop itself. It can be seen in the below image. Can you please suggest way forward!!
![image](https://user-images.githubusercontent.com/74495831/161200750-5e186e7a-95ab-40da-9d7c-e7ce6d479c13.png)

**Reply by HamedHemati** [reply]
@sarat2225 Since the scenario is online, you need to be careful with the buffer part. If you use `ReplyPugin` with the supervised strategy template, it will update the buffer only after each experience. I just skimmed through the paper and it seems like in their setup each experience is a new class, therefore the buffer will be updated after going through all data from each class instead of each batch. Or are you going to implement it in a different way?

**Reply by AntonioCarta** [reply]
@sarat2225 the criterion is independent from the replay buffer. You can change the criterion by passing a function to the strategy or by subclassing and defining your own criterion.
IMO you don't need to do this. By default the strategy is already summing the loss of the replay samples and the new ones. You can make a custom one if you want to balance replay and new losses.

**Comment by ggraffieti** [comment]
Hey @sarat2225, I'm working in the streaming continual learning and I have a working implementation of CBRS as Avalanche __strategy__. 
The code base is still quite dirty and not ready to be published in Avalanche (I'm planning to add CBRS and other streaming methods as soon as the code is well tested and cleaned), but I can share to you (or give you some advice) on CBRS. 
Let me know if you still need help :smiley:

**Reply by sarat2225** [reply]
Hey @ggraffieti Thanks so much for help. I would like to see your code. Kindly share the code to cs19btech11003@iith.ac.in . I am looking forward to share ideas with each other. We also implemented the CBRS using  methods of AGEM strategy and plugin by overriding them. If you are interested in it, I would like to share it with you!!

**Reply by ggraffieti** [reply]
Hey @sarat2225, thinking about that, our code for CBRS is mostly based on the original code that the authors of the paper kindly share with us upon request. Since the main implementation is not done by me (I only inserted their code into an avalanche strategy fixing some minor things) maybe it's better if you directly ask the code to the authors of CBRS. After you have the code if you need help to port the original implementation into avalanche I can give you some hint and advice :grin:

**Reply by AntonioCarta** [reply]
@ggraffieti can you ask the authors if they will allow the integration of their implementation in Avalanche?
